optuna/samplers/_tpe/sampler.py

```diff
@@ -651,3 +782,44 @@ def _get_observation_pairs(
         scores.append(score)
 
     return values, scores
+
+
+def _get_multivariate_observation_pairs(
+    study: Study, param_names: List[str]
+) -> Tuple[Dict[str, List[Optional[float]]], List[Tuple[float, float]]]:
+
+    sign = 1
+    if study.direction == StudyDirection.MAXIMIZE:
+        sign = -1
+
+    scores = []
+    values = {
+        param_name: [] for param_name in param_names
+    }  # type: Dict[str, List[Optional[float]]]
+    for trial in study._storage.get_all_trials(study._study_id, deepcopy=False):
+
+        # We extract score from trial.
+        if trial.state is TrialState.COMPLETE and trial.value is not None:
+            score = (-float("inf"), sign * trial.value)
+        elif trial.state is TrialState.PRUNED:
+            if len(trial.intermediate_values) > 0:
+                step, intermediate_value = max(trial.intermediate_values.items())
+                if math.isnan(intermediate_value):
+                    score = (-step, float("inf"))
+                else:
+                    score = (-step, sign * intermediate_value)
+            else:
+                score = (float("inf"), 0.0)
+        else:
+            continue
+        scores.append(score)
+
+        # We extract param_value from trial.
+        for param_name in param_names:
+            param_value = None
+            if param_name in trial.params:
+                distribution = trial.distributions[param_name]
+                param_value = distribution.to_internal_repr(trial.params[param_name])
```

Comment:
I am not sure if we can make the code faster and clearer by using 2-d `np.ndarray` here. 

In the `_MultivariateParzenEstimator` running behind the sample_relative, sampling, likelihood computation, and transformation of variables are carried out by iterating over parameters. This means we have to use 1-d slices of `np.ndarray` in these methods, and thus I do not expect a much speed up by using 2-d `np.ndarray` here.

As for clarity of using 2-d `np.ndarray`, using `dict` makes it easier to handle samples inside `_MultivariateParzenEstimator` because most of the attributes are stored as `dist`s in `_MultivariateParzenEstimator`. If we use 2-d `np.ndarray`, we have to carefully manage the order of the parameters (for example, by using `OrderedDict` as you pointed out elsewhere).

Reply:
As for using a dense 2d ndarray, I now see that we're clearly looping over each dimension and conditioning the logic based on the distribution. Your implementation looks reasonable. We can however assume that we're always taking the intersection and that all parameters are "not None".

Reply:
OK, I agree that removing the if statement and using assertion here make the code readable.